File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webapi/service/types/ReadContext.java
package org.opcfoundation.webapi.service.types;

import org.eclipse.milo.opcua.stack.core.types.enumerated.TimestampsToReturn;
import org.eclipse.milo.opcua.stack.core.types.structured.ReadValueId;
import org.eclipse.milo.opcua.stack.core.types.structured.RequestHeader;
import org.jspecify.annotations.Nullable;
import org.springframework.web.context.request.NativeWebRequest;

import java.util.List;

public class ReadContext extends ServiceContext {

    final private List<ReadValueId> nodesToRead;
    final private Double maxAge;
    final private TimestampsToReturn timestampsToReturn;

    public ReadContext(
            List<ReadValueId> nodesToRead,
            Double maxAge,
            TimestampsToReturn timestampsToReturn,
            RequestHeader header,
            @Nullable String serverUri,
            NativeWebRequest webRequest)
    {
        super(header,serverUri,webRequest);
        this.nodesToRead = nodesToRead;
        this.maxAge = maxAge;
        this.timestampsToReturn = timestampsToReturn;
    }

    public List<ReadValueId> getNodesToRead() {
        return nodesToRead;
    }

    public Double getMaxAge() {
        return maxAge;
    }

    public TimestampsToReturn getTimestampsToReturn() {
        return timestampsToReturn;
    }
}

File: opcua-webservice-node/src/server/types/contexts/ReadContext.ts
import { ReadRequest, ReadValueId } from "opcua-webapi";
import { ServiceContext } from "./ServiceContext";

export class ReadContext extends ServiceContext
{
    private _nodesToRead: Array<ReadValueId>;
    private _maxAge: number | null;
    private _timestampsToReturn: number | null;
    
    constructor(request: ReadRequest, serverUri?: string)
    {
        super(serverUri, request.RequestHeader);

        this._nodesToRead = (request.NodesToRead) ? request.NodesToRead : [];
        this._maxAge = (request.MaxAge !== undefined) ? request.MaxAge : null;
        this._timestampsToReturn = (request.TimestampsToReturn !== undefined) ? request.TimestampsToReturn : null;
    }

    get nodesToRead() : Array<ReadValueId>
    {
        return this._nodesToRead;
    }

    get maxAge() : number | null
    {
        return this._maxAge;
    }

    get timestampsToReturn() : number | null
    {
        return this._timestampsToReturn;
    }
}
